src/lua/datetime.lua

```diff
@@ -0,0 +1,611 @@
+        -- if there are separate nsec, usec, or msec provided then
+        -- timestamp should be integer
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        elseif fraction ~= 0 then
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+        hms = true
+    end
+
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, offset)
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        y = y or 1970
+        M = M or 1
+        d = d or 1
+        if d < 0 then
+            d = builtin.tnt_dt_days_in_month(y, M)
+        end
+        dt = builtin.tnt_dt_from_ymd(y, M, d)
+    end
+
+    -- .hour, .minute, .second
+    local secs = 0
+    if hms then
+        secs = (h or 0) * 3600 + (m or 0) * 60 + (s or 0)
+    end
+
+    return datetime_new_dt(dt, secs, nsec, offset or 0)
+end
+
+--[[
+    Convert to text datetime values
+
+    - datetime will use ISO-8601 format:
+        1970-01-01T00:00Z
+        2021-08-18T16:57:08.981725+03:00
+]]
+local function datetime_tostring(self)
+    check_date(self, 'datetime.tostring()')
+
+    local buff = ffi.new('char[?]', TOSTRING_BUFSIZE)
+    local len = builtin.datetime_to_string(buff, TOSTRING_BUFSIZE, self)
+    assert(len < TOSTRING_BUFSIZE)
+    return ffi.string(buff)
+end
+
+--[[
+    Dispatch function to create datetime from string or table.
+    Creates default timeobject (pointing to Epoch date) if
+    called without arguments.
+]]
+local function datetime_from(o)
+    if o == nil or type(o) == 'table' then
+        return datetime_new(o)
+    end
+end
+
+--[[
+    Create datetime object representing current time using microseconds
+    platform timer and local timezone information.
+]]
+local function local_now()
+    local d = datetime_new_raw(0, 0, 0)
+    builtin.datetime_now(d)
+    return d
+end
+
+--[[
+    Return table in os.date('*t') format, but with timezone
+    and nanoseconds
+]]
+local function datetime_totable(self)
+    local secs = local_secs(self) -- hour:minute should be in local timezone
+    local dt = local_dt(self)
+
+    return {
+        year = builtin.tnt_dt_year(dt),
+        month = builtin.tnt_dt_month(dt),
+        yday = builtin.tnt_dt_doy(dt),
+        day = builtin.tnt_dt_dom(dt),
+        wday = (ffi.cast('int32_t', builtin.tnt_dt_dow(dt)) + 1) % 7,
+        hour = math_floor((secs / 3600) % 24),
+        min = math_floor((secs / 60) % 60),
+        sec = secs % 60,
+        isdst = false,
+        nsec = self.nsec,
+        tzoffset = self.tzoffset,
+    }
+end
+
+local function datetime_update_dt(self, dt, new_offset)
+    local epoch = local_secs(self)
+    local secs_day = epoch % SECS_PER_DAY
+    epoch = (builtin.tnt_dt_rdn(dt) - DAYS_EPOCH_OFFSET) * SECS_PER_DAY +
+            secs_day
```

Comment:
What is happening here? I don't understand. The same for the other 'update' functions.

Reply:
Ok, let me try to explain it a bit:

datetime_ymd_update/datetime_update_dt
============================

* c-dt uses dt_t (32-bit integer) type to represent a number days since Rata Die date (0001-01-01);
* for given values of `y` (year), `M` (month), `d` (day) we calculate corresponding dt value using `dt_from_ymd()` call;
* so the trick is to split stored value in `epoch` field to `dt` (number of days) and the rest of seconds since beginning of local day (`secs_day`);
* so the new epoch will be constructed as function of new `dt` (as beginning of day) + the rest of seconds we used to have in time before - `dt * SECS_PER_DAY + secs_day`.
* the only left is shift we need to proceed for our coordinates: epoch is counted since 1970-01-01, not `0001-01-01`, this this subtraction of `DAYS_EPOCH_OFFSET` (719163)

Sorry for the confusion created with unnecessary call to `builtin.tnt_dt_rdn` that was harmless in our build (as it was not changing input value), but excessive semantically.

datetime_hms_update
==============

* first we add timezone shift, to make sure that hour, minutes and seconds will be local in our arithmetcis - `epoch = local_secs(self)`;
* we get rid of all seconds since beginning of day (removing hour, minutes and seconds), calculating value of seconds of midnight for this day  - `secs_day = epoch - (epoch % SECS_PER_DAY)`;
* then new epoch will be addition of former midnight second + h:m:s passed - `secs_day + h * 3600 + m * 60 + s`;
* and this seconds will be unshift back (normalized to UTC) given `new_offset` - `self.epoch = utc_secs(secs_day + h * 3600 + m * 60 + s, new_offset)`

Reply:
Hope, this is clearer now :)